internal/resource.go

```diff
@@ -1,140 +1,263 @@
 	if len(urls) < 1 {
 		return nil, fmt.Errorf("empty url list")
 	}
-	url := urls[0]
 	ctx := context.Background()
-	path, err := GetUrltoTempFile(url, ctx)
+
+	path, err := GetUrltoTempFile(urls[0], ctx)
 	if err != nil {
-		return nil, fmt.Errorf("failed to get url: %s", err)
+		return nil, err
 	}
 	defer os.Remove(path)
+
 	integrity, err := getIntegrityFromFile(path, algo)
 	if err != nil {
-		return nil, fmt.Errorf("failed to compute ressource integrity: %s", err)
+		return nil, err
 	}
-	return &Resource{Urls: urls, Integrity: integrity, Tags: tags, Filename: filename}, nil
+
+	return &Resource{
+		Urls:      urls,
+		Integrity: integrity,
+		Tags:      tags,
+		Filename:  filename,
+	}, nil
 }
 
-// getUrl downloads the given resource and returns the path to it.
 func getUrl(u string, fileName string, ctx context.Context) (string, error) {
-	_, err := url.Parse(u)
-	if err != nil {
-		return "", fmt.Errorf("invalid url '%s': %s", u, err)
+	if _, err := url.Parse(u); err != nil {
+		return "", fmt.Errorf("failed to download '%s': invalid URL", u)
 	}
-	log.Debug().Str("URL", u).Msg("Downloading")
-	err = requests.
-		URL(u).
+
+	req := requests.URL(u).
 		Header("Accept", "*/*").
-		ToFile(fileName).
-		Fetch(ctx)
+		ToFile(fileName)
+
+	err := req.Fetch(ctx)
 	if err != nil {
-		return "", fmt.Errorf("failed to download '%s': %s", u, err)
+		return "", fmt.Errorf("failed to download '%s': %v", u, err)
 	}
-	log.Debug().Str("URL", u).Msg("Downloaded")
+	log.Debug().Msg("Download completed")
 	return fileName, nil
 }
 
-// GetUrlToDir downloads the given resource to the given directory and returns the path to it.
 func GetUrlToDir(u string, targetDir string, ctx context.Context) (string, error) {
-	// create temporary name in the target directory.
 	h := sha256.New()
 	h.Write([]byte(u))
 	fileName := filepath.Join(targetDir, fmt.Sprintf(".%s", hex.EncodeToString(h.Sum(nil))))
 	return getUrl(u, fileName, ctx)
 }
 
-// GetUrlWithDir downloads the given resource to a temporary file and returns the path to it.
 func GetUrltoTempFile(u string, ctx context.Context) (string, error) {
 	file, err := os.CreateTemp("", "prefix")
 	if err != nil {
-		log.Fatal().Err(err)
+		return "", err
 	}
-	fileName := file.Name()
-	return getUrl(u, fileName, ctx)
+	return getUrl(u, file.Name(), ctx)
+}
+
+func GetFileHash(filePath string) (string, error) {
+	data, err := ioutil.ReadFile(filePath)
+	if err != nil {
+		return "", fmt.Errorf("failed to read file: %w", err)
+	}
+
+	hasher := sha256.New()
+	hasher.Write(data)
+	return hex.EncodeToString(hasher.Sum(nil)), nil
+}
+
+func (r *Resource) Contains(url string) bool {
+	for _, u := range r.Urls {
+		if u == url {
+			return true
+		}
+	}
+	return false
+}
+
+func (r *Resource) Download(dir string, mode os.FileMode, ctx context.Context) error {
+	// Try cache first if available
+	if r.CacheUri != "" {
+		token := os.Getenv("GRABIT_ARTIFACTORY_TOKEN")
```

Comment:
You're repeating the env var in too many places, please make it a constant and use the const everywhere.